src/passes/AddThrows.cpp

```diff
@@ -0,0 +1,93 @@
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+struct AddThrowsPass : WalkerPass<PostWalker<AddThrowsPass>> {
+
+  void visitBinary(Binary* curr) {
+    switch (curr->op) {
+      case DivSInt32:
+      case DivUInt32:
+      case RemSInt32:
+      case RemUInt32:
+      case DivSInt64:
+      case DivUInt64:
+      case RemSInt64:
+      case RemUInt64: {
+        Builder builder(*getModule());
+        auto type = curr->right->type;
+        auto eqz = type == Type::i32 ? EqZInt32 : EqZInt64;
+        auto scratch = builder.addVar(getFunction(), type);
+        curr->right = builder.makeIf(
+          builder.makeUnary(eqz,
+                            builder.makeLocalTee(scratch, curr->right, type)),
+          builder.makeThrow("DivByZero", {}),
+          builder.makeLocalGet(scratch, type),
+          type);
+        return;
+      }
+      default:
+        return;
+    }
+  }
+
+  template<typename T> std::optional<Index> handleRef(T* curr) {
+    Builder builder(*getModule());
+    auto type = curr->ref->type;
+    if (type.isNonNullable()) {
+      return std::nullopt;
+    }
+    auto scratch = builder.addVar(getFunction(), type);
+    curr->ref = builder.makeIf(
+      builder.makeRefIsNull(builder.makeLocalTee(scratch, curr->ref, type)),
+      builder.makeThrow("NullDeref", {}),
+      builder.makeLocalGet(scratch, type),
+      type);
+    return scratch;
+  }
+
+  template<typename T>
+  void handleIndex(T* curr, std::optional<Index> refLocal) {
+    Builder builder(*getModule());
+    if (!refLocal) {
+      refLocal = builder.addVar(getFunction(), curr->ref->type);
+      curr->ref = builder.makeLocalTee(*refLocal, curr->ref, curr->ref->type);
+    }
+    auto scratch = builder.addVar(getFunction(), Type::i32);
+    curr->index = builder.makeIf(
```

Comment:
I have been using `br_if` since inserting a conditional makes a lot of optimizations (like Local CSE) less effective.

Reply:
Interesting! I think wasm-opt generally prefers to emit conditionals because they end up being fewer bytes than a block + br_if. I don't have an intuition for why a conditional would be worse for CSE. Can you explain that?

Reply:
As far as I understand, CSE currently does not work across basic blocks, except when they are adjacent. The expressions before and after a conditional belong to different non-adjacent basic blocks, while the basic blocks before and after a `br_if` are considered adjacent. See file `ir/linear-execution.h`.

I'm doing a non-local transformation, actually: I'm putting the target of the `br_if` at the end of the current function (or at the end of a `try` block).